### PR TITLE
Add WindowBuilder::with_outer_position

### PR DESCRIPTION
**Pull request by xixixao**
- [x] Tested on all platforms changed
- [x] Compilation warnings were addressed
- [x] `cargo fmt` has been run on this branch
- [x] `cargo doc` builds successfully
- [x] Added an entry to `CHANGELOG.md` if knowledge of this change could be valuable to users
- [x] Updated documentation to reflect any user-facing changes, including notes of platform-specific behavior
- [x] Created or updated an example program if it would help users understand this functionality
- [x] Updated [feature matrix](https://github.com/rust-windowing/winit/blob/master/FEATURES.md), if new features were added or implemented

I think this finishes the important platforms and we could actually merge the branch to master?

cc @murarth @Osspial

### CODE REVIEW

**Comment by xixixao** [comment]
I combined all the changes into a single commit on top of master. Let me know if I should first update this branch, or if I should open PR against master.

**Comment by LoganDark** [comment]
Why does this simple PR have 153 files changed? It's impossible for anyone to sort through this.

**Comment by maroider** [comment]
Could you make this PR against `master` instead, since this seems to finish the `with_outer_position` feature?
This PR would then supplant #1240/#1239.
You may or may not also want to add @Osspial and @murarth as co-authors, if you're feeling polite.

**Comment by xixixao** [comment]
Changed base to master, added contributors, fixed X11 (I can't `check` it so I have to wait for CI to tell me whether it's broken - let me know if there is a better way).

**Comment by maroider** [comment]
It seems like GitHub won't accept your co-author assignments. Try this instead:
```
Co-authored-by: Osspial <osspial@gmail.com>
Co-authored-by: Murarth <murarth@gmail.com>
```

PS: Osspial is no longer a mainainer for Windows. @msiglreith is the one you want for now.

File: src/platform_impl/macos/window.rs
Comment (ArturKovacs) [inline src/platform_impl/macos/window.rs:1223]: On top of the change in the wording of the documentation, I would suggest moving this function to `util/mod.rs` right underneath `bottom_left_to_top_left` (Which really doesn't have a good name or description in my opinion, but you don't need to be concerned about that.)

**Review by ArturKovacs — CHANGES_REQUESTED** [review]

File: src/platform_impl/macos/window.rs
Comment (xixixao) [reply src/platform_impl/macos/window.rs:178]: This is unlikely to be solvable. Creating window and then setting its position break the whole point of this feature, which is to avoid the flickering when creating a window and then repositioning it.

I'm also not sure which position is wrong one, whether this one or the one in set_outer_position.

Comment (xixixao) [reply src/platform_impl/macos/window.rs:178]: Guessing or hardcoding the title bar height is unlikely to work, due to macOS UI changes like this:

https://www.reddit.com/r/MacOSBeta/comments/hq00ll/insane_titlebar_height_and_question_on_logical/

I think we'll have to settle on documenting that this method sets "inner" position on MacOS, despite its name, due to platform limitations.

**Comment by xixixao** [comment]
Ready for re-review fyi @ArturKovacs @msiglreith

**Review by msiglreith — APPROVED** [review]
Thanks, lgtm (Windows).

The default Win10 theme has some invisible borders which result in some minor offsets (7/8pts) of the visibile position. But not too important imo.

File: src/platform_impl/macos/window.rs
Comment (ArturKovacs) [reply src/platform_impl/macos/window.rs]: Yep we should definitely not hardcode or guess the menubar height when the function name promises to set the outer position. For the same reason I don't think we should just set the inner position when the function is called `with_outer_position`. But I agree that it would be nice to have at least some way to set the startup position. So at this point the only thing I have problem with is the function name. I think calling it `with_approximate_position` or something similar would be completely fine. And the documentation should of course describe what this means for each platform.

Comment (ArturKovacs) [reply src/platform_impl/macos/window.rs]: What do you think @msiglreith @murarth?

**Comment by xixixao** [comment]
This should be ready for merging btw. cc @ArturKovacs

File: src/window.rs
Comment (ArturKovacs) [inline src/window.rs:132]: This should specify what kind of position it is, and to me, the wording is a bit hard to understand. I suggest

"The outer position. In other words, the top left corner of the title-bar. There may be a small gap between this position and the window due to specifics of the Window Manager"

File: src/window.rs
Comment (ArturKovacs) [inline src/window.rs:125]: The Platform-specific list should be formatted as a list, that is each item beggining with a `-` like this:
```
- **macOS**: ...
- **Window**: ...
```

Also under Platform-specific, you should describe the kind of position this means on all of the platforms. For example
```
- **X11**: The outer position.
- **Others**: This is ignored.
```

**Comment by ArturKovacs** [comment]
I know I'm being unusually picky, but it's mostly because this one functionality does different things on different platforms, so I just want to make sure that whoever comes across this will have no confusion about what it does.

Furthermore considering that the name of a public function and a public field has changed, I believe @msiglreith and @murarth need to approve this before it can be merged.

**Comment by maroider** [comment]
>Furthermore considering that the name of a public function and a public field has changed

Isn't this PR purely additive when compared against master?

**Comment by ArturKovacs** [comment]
What I meant was that there were changes in the API since those maintainers approved therefore their approval doesn't apply to the current state.

File: src/window.rs
Comment (msiglreith) [reply src/window.rs:132]: This is not yet addressed

**Comment by msiglreith** [comment]
Fine for me once the latest comments by @ArturKovacs have been addressed.

**Comment by xixixao** [comment]
Addressed @ArturKovacs's comments.

**Review by murarth — APPROVED** [review]

**Review by ArturKovacs — APPROVED** [review]

**Pull request merged**
